### PR TITLE
Ch refactor code

### PR DESCRIPTION
**Pull request by ibetween**
### What does this PR do?
This PR contains flask restful api for the **Thriller diary**
### Description
The following features has been added:

- Fetching of all entries
- Fetching of a single entry
- Editing of a single entry
- Updating/Modifying a single entry.
-  Deleting of a single entry
- fetching of the profile page
- changing of password
### status
[Build Status](https://travis-ci.org/james-chege/thriller_diary_api)
[Coverage Status](https://coveralls.io/github/james-chege/thriller_diary_api?branch=develop)
[Codacy Badge](https://www.codacy.com/app/james-chege/thriller_diary_api?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=james-chege/thriller_diary_api&amp;utm_campaign=Badge_Grade)

### How to access the project locally?
Fork this repository to your github account and clone from there(_NB: clone from your github account - after forking_).This will help you to contribute to this project.

[Create a python Virtual environment and Activate it](https://virtualenv.pypa.io/en/stable/). A virtual environment is effective when working on multiple projects. Each project will have its own development enviroment.

__Install Dependencies__(_Note: This should be done in the created virtual environment_)
```py
 pip install -r requirements.txt
```
__Set environment variable__
```py
 export APP_SETTINGS="development"
```

# setup Database
* make sure you have [__postgresql__](https://www.postgresql.org/download/linux/ubuntu/) installed and running properly.

* run the following command to create the database:
* 
```sql
- psql -U postgres

- CREATE DATABASE thriller;
```

* Set the database configuration variables.

```
export DATABASE="thriller"
export PASSWORD="your postgres user password"
```

On the root folder of this project run __`python create_db.py`__ to create database for the Api.

__Start Server__
```py
python run.py
```

[__Use postman app to send request to app.__](https://www.getpostman.com/)
### Endpoints

The following is a list of available endpoints in this application

|EndPoint               | Functionality|
| ------------------------------------ | ------------------------ |
|POST api/v1/auth/signup    |Register a user|
|POST api/v1/auth/login |Login a user|
|GET api/v1/entries |fetch all entries|
|GET api/v1/entries/<int:entryId> |fetch a single entry|
|POST api/v1/entries |Add an entry|
|PUT api/v1/entries/<int:entryId> |modify an entry|
|DELETE api/v1/entries/<int:entryId> |Deletes a user's entry|
|GET api/v1/users/profile/ |Retrieves logged in user's info|
|PUT api/v1/users/change_password/ |Modifies user's password|

# Testing
* __Creating database for testing__
To create database for testing, run __`python create_test_db.py`__.

`py.test --cov=app`

* change the database name to tests
`export DATABASE="tests"`

### CODE REVIEW

File: app/db.py [hunk 1]
```diff
@@ -1,11 +1,12 @@
 import psycopg2
 import os
 class Connection(object):
+    """ creates connection to the database """
 
     def __init__(self):
         # ########### OFFICIAL DB CREDENTIALS ##############
-        self.HOSTNAME = 'localhost'
-        self.USERNAME = 'postgres'
+        self.HOSTNAME = 'ec2-50-16-241-91.compute-1.amazonaws.com'
```
Comment (NgangaMaryanne) [inline app/db.py]: Get this from the environment variable too. It is especially dangerous to commit database credentials to Git.

File: app/db.py [hunk 2]
```diff
@@ -1,11 +1,12 @@
...
+        self.USERNAME = 'tkvjvlrggrkghq'
```
Comment (NgangaMaryanne) [inline app/db.py]: See above comment.

Comment (ibetween) [reply app/db.py]: I will rectify it right now. Thanks

File: app/models.py [hunk 3]
```diff
@@ -60,3 +125,46 @@ def decorated(*args, **kwargs):
...
+        cur.execute(query, data)
+        self.db.commit()
+        response = {"status": "success", 
+            "Registered": {"Email":str(user_email), 
+                            "Username":str(username)}}
+        return response, 201
+    
+    def login_user(self, user_email, user_password):
+        checker = self.db.cursor()
+        checker.execute("SELECT * FROM users")
```
Comment (NgangaMaryanne) [inline app/models.py:159]: You could just get the specific user from the database, Its just added latency on the database when you get all users just so you could filter for one user. It is a very expensive operation on the database side. Use the WHERE clause to get just the user you need. https://www.w3schools.com/sql/sql_where.asp

Comment (ibetween) [reply app/models.py:159]: That is much easier thanks you so much :100:

File: tests/test_auth.py [hunk 4]
```diff
@@ -19,54 +19,71 @@ def setUp(self):
...
     def test_registration_success(self):
+        """ test registration success """
         self.register = {"username":"spinderman", "password":"spinderman", "email":"spinderman@gmail.com"}
         req = self.client().post(self.register_route, data=self.register) # pylint: disable=unused-variable
         self.assertEqual(req.status_code, 201)
+        self.assertIn('success', str(req.data))
 
     def test_login_success(self):
+        """ test login success """
         self.user_data = {"email":"thor@gmail.com", "password":"thor"}
```
Comment (NgangaMaryanne) [inline tests/test_auth.py]: I do not see where this person was registered before they could login. So how is this person able to login?

File: tests/test_auth.py [hunk 5]
```diff
@@ -19,54 +19,71 @@ def setUp(self):
...
-    def test_invalid_input(self):
-        self.user_data_empty = {"email":""}
+    def test_invalid_username(self):
+        """ test for invalid username """
         self.user_data_invalid_username = {"username":"#$superman", "email":"superman@gmail.com", "password":"69mansuper"}
+        req3 = self.client().post(self.register_route, data=self.user_data_invalid_username)
+        self.assertEqual(req3.status_code, 401)
+        self.assertIn('Invalid username', str(req3.data))
+
+    def test_invalid_password(self):
```
Comment (NgangaMaryanne) [inline tests/test_auth.py]: Break down this test into smaller test suites, each testing just one single thing.

File: tests/test_auth.py [hunk 6]
```diff
@@ -19,54 +19,71 @@ def setUp(self):
...
-        self.assertEqual(req2.status_code, 401)
-        self.assertIn('Invalid email', str(req2.data))
-        self.assertEqual(req3.status_code, 401)
-        self.assertIn('Invalid username', str(req3.data))
         self.assertEqual(req.status_code, 401)
         self.assertIn('Check your details', str(req.data))
-        self.assertEqual(req5.status_code, 401)
-        self.assertIn('Oops!', str(req5.data))
+        
+    def test_invalid_email_inputs(self):
```
Comment (NgangaMaryanne) [inline tests/test_auth.py:83]: See above comment

File: tests/test_entries.py [hunk 7]
```diff
@@ -152,18 +152,22 @@ def test_landing_page_message(self):
...
         req = self.client().get('/api/v1/entries', headers={"access-token":"invalid_access_token"})
         self.assertEqual(req.status_code, 401)
+        self.assertIn('token is invalid', str(req.data))
     def test_empty_token(self):
         """ test empty token"""
         req = self.client().get('/api/v1/entries', headers={"access-token":None})
         self.assertEqual(req.status_code, 401)
+        self.assertIn('Provided token is invalid', str(req.data))
 
     def test_fetch_single_entry(self):
```
Comment (NgangaMaryanne) [inline tests/test_entries.py:173]: See above comment

File: tests/test_auth.py (hunk 4 above)
Comment (ibetween) [reply tests/test_auth.py]: let me fix that right now

File: tests/test_auth.py (hunk 5 above)
Comment (ibetween) [reply tests/test_auth.py]: Alright. Noted with thanks

**Pull request merged**